jcl/src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -855,37 +856,35 @@ public final void join(long timeoutInMilliseconds) throws InterruptedException {
  * @see			java.lang.ThreadDeath
  */
 public final synchronized void join(long timeoutInMilliseconds, int nanos) throws InterruptedException {
-	if (timeoutInMilliseconds < 0 || nanos < 0 || nanos > NANOS_MAX)
+	if ((timeoutInMilliseconds < 0) || (nanos < 0) || (nanos > NANOS_MAX)) {
 		throw new IllegalArgumentException();
+	}
 
-	if (!started || isDead()) return;
-
-	// No nanosecond precision for now, we would need something like 'currentTimenanos'
+	if (!started || isDead()) {
+		return;
+	}
 
-	long totalWaited = 0;
 	long toWait = timeoutInMilliseconds;
 	boolean timedOut = false;
 
 	/*[PR 1PQM757] Even though we do not have nano precision, we cannot wait(0) when any one of the parameters is not zero */
-	if (timeoutInMilliseconds == 0 & nanos > 0) {
+	if ((timeoutInMilliseconds == 0) && (nanos > 0)) {
 		// We either round up (1 millisecond) or down (no need to wait, just return)
-		if (nanos < 500000)
+		if (nanos < 500000) {
 			timedOut = true;
-		else
+		} else {
 			toWait = 1;
+		}
 	}
 	/*[PR 1FJMO7Q] A Thread can be !isAlive() and still be in its ThreadGroup. Use isDead() */
 	while (!timedOut && !isDead()) {
-		long start = System.currentTimeMillis();
+		final long start = System.nanoTime();
 		wait(toWait);
```

Comment:
> If both arguments are zero, this should wait indefinitely; see [1].

It does invoke `wait(toWait)` which is `zero`, i.e., wait indefinitely. However there is a problem here indeed, if the thread is notified, it need check `isDead()` before finishing the wait. This is an existing problem, will fix it.

> This rounding to milliseconds should be removed and wait(long, long) used directly.

There is a drawback to use `wait(long, long)`, the `toWait` has to be calculated after each `wait(toWait)` whenever the thread is notified. RI uses similar strategy rounding to `ms`.

> The new version could, theoretically, wait much longer than desired if many calls to wait() takes less than half a millisecond so that in each case, toWait is not reduced.

There is no change, in such case,  `timedOut` is set to `true`, `join` returns immediately.

Reply:
I take back my concern when both parameter values are zero: that case was and still is handled correctly.

The current version doesn't suffer from the truncation introduced here. `currentTimeMillis()` will eventually return a larger value and `totalWaited` would be updated accordingly. The new version does not guarantee progress.

It seems silly to ignore the provided nanoseconds by calling `wait(long)` which in turn calls `wait(long,long)`; this should call the latter directly with the extra information available here.

Reply:
Use `TimeUnit.NANOSECONDS.timedWait(this, toWaitNano)` which invokes `wait(long,long)` later.